### PR TITLE
Feature/array higher order functions Daniel Toledo

### PR DESCRIPTION
**Pull request by danieltolmon**

### CODE REVIEW

File: staff/daniel-toledo/array-higher-order-functions/filter.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * 
+ * @throws {TypeError} - when array is not an array
+ * @throws {TypeError} - when callback is not a function
+ * 
+ * @return {Array} - new array with filtered results
+ */
+
+function filter(array, callback) {
+
+    if (!(array instanceof Array)) throw TypeError(array + ' should be an error');
```
Comment (maryammdot) [inline staff/daniel-toledo/array-higher-order-functions/filter.js]: the error message is not really accurate jeje

File: staff/daniel-toledo/array-higher-order-functions/filter.test.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    var arr=[1,2,3,4,5];
+
+    res=filter(arr,function(el) {return el%2===0});
+
+    assert(res.toString()===[2,4].toString(),'Unexpected value');
+    assert(arr.toString() === [1,2,3,4,5].toString(),'Input array should not change')
+
+});
+
+test('returns array when function does not find', function(){
```
Comment (maryammdot) [inline staff/daniel-toledo/array-higher-order-functions/filter.test.js]: You could be more accurate in the test title, so it should return empty array...

File: staff/daniel-toledo/array-higher-order-functions/filter.test.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    var arr=[1,2,3,4,5];
+
+    res=filter(arr,function(el) {return el>10});
+
+    assert(res.toString()===[].toString(),'Unexpected value');
+    assert(arr.toString() === [1,2,3,4,5].toString(),'Input array should not change')
+
+});
+
+test('array is not an Array', function(){
```
Comment (maryammdot) [inline staff/daniel-toledo/array-higher-order-functions/filter.test.js]: message could be it should fail...

File: staff/daniel-toledo/array-higher-order-functions/filter.js (hunk 1 above)
Comment (danieltolmon) [reply staff/daniel-toledo/array-higher-order-functions/filter.js]: done!!

File: staff/daniel-toledo/array-higher-order-functions/filter.test.js (hunk 2 above)
Comment (danieltolmon) [reply staff/daniel-toledo/array-higher-order-functions/filter.test.js]: `test('it should return an array when the callback does not find any value', function(){`
done!!

File: staff/daniel-toledo/array-higher-order-functions/filter.test.js (hunk 3 above)
Comment (danieltolmon) [reply staff/daniel-toledo/array-higher-order-functions/filter.test.js]: `test('it should throw an error when array is not an Array'`
done!!
